python/amici/sbml_utils.py

```diff
@@ -0,0 +1,600 @@
+        expr, *, replace_time: bool = False, pretty: bool = False, **settings
+        ) -> str:
+    """
+    Prints a SymPy expression to a MathML expression parsable by libSBML.
+
+    :param expr:
+        expression to be converted to MathML (will be sympified).
+
+    :param replace_time:
+        replace the AMICI time symbol with the SBML time symbol.
+
+    :param pretty:
+        prettify the resulting MathML.
+    """
+    with evaluate(False):
+        expr = sp.sympify(expr)
+        if replace_time:
+            expr = expr.subs(amici_time_symbol, sbml_time_symbol)
+    return MathMLSbmlPrinter(settings).doprint(expr, pretty=pretty)
+
+
+def sbmlMathAST(expr, **kwargs) -> libsbml.ASTNode:
+    """
+    Convert a SymPy expression to SBML math AST.
+
+        :param expr:
+            expression to be converted (will be sympified).
+
+        :param kwargs:
+            extra options for MathML conversion.
+    """
+    mathml = sbmlMathML(expr, **kwargs)
+    ast = libsbml.readMathMLFromString(mathml)
+    if ast is None:
+        raise SbmlException(
+            f'error while converting the following expression to SBML AST.\n'
+            f'expression:\n{expr}\n'
+            f'MathML:\n{pretty_xml(mathml)}'
+        )
+    return ast
+
+
+def setSbmlMath(obj: libsbml.SBase, expr, **kwargs) -> None:
+    """
+    Set the math attribute of a SBML node using a SymPy expression.
+
+    :param obj:
+        SBML node supporting `setMath` method.
+
+    :param expr:
+        expression to which the math attribute of `obj` should be se to
+        (will be sympified).
+
+    :param kwargs:
+        extra options for MathML conversion.
+    """
+    mathml = sbmlMathAST(expr, **kwargs)
+    if obj.setMath(mathml) != libsbml.LIBSBML_OPERATION_SUCCESS:
+        raise SbmlException(
+            f'Could not set math attribute of SBML object {obj}\n'
+            f'expression:\n{expr}\n'
+            f'MathML:\n{pretty_xml(mathml)}'
+        )
+
+
+###############################################################################
+# MathML to Sympy conversion
+
+
+def mathml2sympy(
+        mathml: str, *, evaluate: bool = False, locals=None,
+        expression_type=None
+        ):
+    ast = libsbml.readMathMLFromString(mathml)
+    if ast is None:
+        raise ValueError(
+            f'libSBML could not parse MathML string:\n{pretty_xml(mathml)}'
+        )
+
+    formula = _parse_logical_operators(libsbml.formulaToL3String(ast))
+
+    if evaluate:
+        expr = sp.sympify(formula, locals=locals)
+    else:
+        with sp.core.parameters.evaluate(False):
+            expr = sp.sympify(formula, locals=locals)
+
+    expr = _parse_special_functions(expr)
+
+    if expression_type is not None:
+        _check_unsupported_functions(expr, expression_type)
+
+    return expr
+
+
+def _check_unsupported_functions(
+        sym: sp.Basic,
+        expression_type: str,
+        full_sym: sp.Basic = None
+        ):
+    """
```

Comment:
This looks like a copy of `import_utils._check_unsupported_functions`. Please check which one is newer.